opendrift/readers/basereader/unstructured.py

```diff
@@ -0,0 +1,194 @@
+import logging
+logger = logging.getLogger(__name__)
+
+from .variables import Variables
+
+
+class UnstructuredReader(Variables):
+    """
+    An unstructured reader. Data is gridded irregularily.
+
+    In order to support the heterogeneous types of unstructured grids, the
+    unstructured readers are built up slightly differently than the
+    :class:`.structured.StructuredReader` readers. This class, which readers of
+    unstrctured grids subclass, provide several methods for aiding in
+    interpolation and caching.
+
+    The initial type of grid that this class supports are `triangular prisms <https://en.wikipedia.org/wiki/Types_of_mesh#Triangular_prism>`_. Unstructured in xy-coordinates, x and y is constant in z. z might be non-cartesian (e.g. sigma-levels).
+
+    Caching using UnstructuredBlock
+    -------------------------------
+
+    TODO:
+
+    The `StructuredReader`s return a 2D field for `meshgrid(x,y)`. This is not
+    so meaningful for `UnstructuredReader` since the variables are iregularily
+    placed. The `get_variables` method should therefore be a method required by
+    the `StructuredReader` and `ContinuousReader`, and maybe renamed to
+    `get_block` for `StructuredReader`.
+
+    For the `UnstructuredReader` we need the equivalent of `get_block`, maybe
+    `get_subset` which sets up an `UnstructuredBlock` for the given area or
+    volume. This probably requires :meth:`_get_variables_impl_`/ :meth:`__get_variables_derived__`
+    to be moved to the outside of :meth:`_get_variables_interpolated_`.
+
+    """
+
+    var_block_before = None
+    var_block_after = None
+    boundary = None
+
+    # nodes
+    x = None
+    y = None
+    node_variables = None # list of std-name variables defined at nodes
+    nodes_idx = None
+
+    # faces
+    xc = None
+    yc = None
+    face_variables = None # list of std-name variables defined at center of faces
+    faces_idx = None
+
+    def __init__(self):
+        super().__init__()
+
+        # Dictionaries to store blocks of data for reuse (buffering)
+        self.var_block_before = {}  # Data for last timestep before present
+        self.var_block_after = {}  # Data for first timestep after present
+
+    def _get_variables_interpolated_(self, variables, profiles,
+                                   profiles_depth, time,
+                                   reader_x, reader_y, z):
+
+        env = self._get_variables_impl_(variables,
+                                      profiles,
+                                      profiles_depth,
+                                      time,
+                                      reader_x,
+                                      reader_y,
+                                      z)
+
+        # We probably have to use an UnstructuredBlock to store closest time-steps, and thus avoid fetching
+        # more data on every call.
+
+        logger.debug('Fetched env-before')
+        env_profiles = None
+        if profiles is not None:
```

Comment:
This is presumably a temporary fix?

Reply:
It was probably copied from the code that previously handled analytical readers.